# Assignment2 - Supervised Machine Learning Fundamentals

**Name:** Edward Huang  
**NetID:** yh475  
**Collaborators:** N/A


---
## Exercise 1 - Conceptual Questions on Supervised Learning I

**1.1**     
Expect better.          
Because large n can fit the underlying pattern more accuratly without as much risk of overfitting. Also, large n provides enough information to distinguish the singal from the noise, allowing a flexible model to find a closer fit to teh true functional form.

**1.2**     
Expect worse.       
Few data means the flexible model will find the pattern in the random noise, which is overfitting. Inflexible model perform better because their contrains prevent them from chasing every outlier in a small sample.

**1.3**     
Expect better.          
The flexible method makes strong assumptions about the shape of the data. If the true relationship is curvy or complex, and inflexible model will have high bias.

**1.4**     
Expect worse.           
High variance means the data has high noise. A flexible method will follow the noise, mistaking random fluctuations for the actual relationship. Inflexible method is more robust because it effectively averages out the noise to be more simpler.


---
## Exercise 2 - Conceptual Questions on Supervised learning II

**2.1**     
(i) Regression, because the response variable is the salary and salary is the continuous value.     
(ii) Inference, becuae the question wants us to understand which factors affect CEO salary, which means we need to know the relationship between different variable, not just predict the new data.         
(iii) Sample n = 500, predictor p = 3(profit, number of employees, industry)        

**2.2**         
(i) Classification, because the response variable is whether the product is a success or failure, which is discrete             
(ii) Prediction, because the goal is to determine the likely outcome a future product based on existing data.           
(iii) Sample n = 20, predictor p = 13

**2.3**

(i) Regression, because the response variable is % change in the US dollar, which is a continuous numerical value.          
(ii) Prediction, because we're interested in using world market changes to estimate the future movement of dollar.          
(iii) Sample n = 52 (52 weeks per year), predictor p = 3 (% change in US, British, German)

---
## Exercise 3 - Classification using KNN

**3.1**

In [3]:
import numpy as np
import pandas as pd
X = np.array([[ 0, 3, 0],
              [ 2, 0, 0],
              [ 0, 1, 3],
              [ 0, 1, 2],
              [-1, 0, 1],
              [ 1, 1, 1]])
y = np.array(['r','r','r','b','b','r'])

test_point = np.array([0, 0, 0])
distance = np.linalg.norm(X - test_point, axis = 1)
df = pd.DataFrame({'Obs.': np.arange(1, 7),'x1': X[:, 0],'x2': X[:, 1],'x3': X[:, 2],'y': y,'Distance': distance})

print(df.to_string(index = False))

 Obs.  x1  x2  x3 y  Distance
    1   0   3   0 r  3.000000
    2   2   0   0 r  2.000000
    3   0   1   3 r  3.162278
    4   0   1   2 b  2.236068
    5  -1   0   1 b  1.414214
    6   1   1   1 r  1.732051


**3.2**             
Predict when k = 1          
Ans: Blue, when k = 1, the prediction is based on the single observation in the training set that is closest to the test point, which is obs 5 (blue).

**3.3**         
Predict when k = 3      
Ans: Red        
Base on the Euclidean distance, the three observations closest to the test point are obs5, obs6, obs2.      
2 red labels and 1 blue label, so KNN predicts y = red.

**3.4**         
I would expect the best value of k to be small.         
A small k provides a more flexible model with a lower bias, allowing the decision boundary to follow non-linear patterns in the data.       
Also if K increases, the decision boundary becomes more linear because it averages the labels over a larger neighborhood.
Therefore, a highly non-linear Bayes ecision boundary requires the model to capture complex shapes, a small K is necessary to provide the required.